docs: miscellaneous improvements (#3219)
docker-mailserver/docker-mailserver: docs/content/usage.md

title: Usage
---

This pages explains how to get started with DMS. The guide uses Docker Compose as a reference. In our examples, a volume mounts the host location [`docker-data/dms/config/`][docs-dms-config-volume] to `/tmp/docker-mailserver/` inside the container.

> [docs-dms-config-volume]: ./config/advanced/optional-config/

Reviewer: ```suggestion
[docs-dms-config-volume]: ./config/advanced/optional-config.md
```

| `push` on `master`      | `edge`                        |
| `push` a tag (`v1.2.3`) | `1.2.3`, `1.2`, `1`, `latest` |

> [CI/CD]: https://github.com/docker-mailserver/docker-mailserver/actions

Author: > **NOTE:** Slight consistency issue in the docs with this convention. Most are using `[github-*]:` but some are using `[gh-*]:`.

Can be resolved with a follow-up PR :)

Issue the following commands to acquire the necessary files:

``` BASH
TAG='v11.3.1'
DMS_GITHUB_URL="https://github.com/docker-mailserver/docker-mailserver/blob/${TAG}"
wget "${DMS_GITHUB_URL}/docker-compose.yml"
wget "${DMS_GITHUB_URL}/mailserver.env"
```

> Then edit `docker-compose.yml` to your liking; substitute `mail.example.com` according to your FQDN, and if you want to use SELinux for the `./docker-data/dms/config/:/tmp/docker-mailserver/` mount, append `-z` or `-Z`. Then configure the environment specific to the mail server by editing [`mailserver.env`][docs-environment], but keep in mind that:

Author: I collapsed this because point 1 was never suited to be there anyway and I felt it was more concise to just write it as text and not with an enumeration.

I can bring back the enumeration though, if you like it more, I will not be insisting on using the sentences.

> Then edit `docker-compose.yml` to your liking; substitute `mail.example.com` according to your FQDN, and if you want to use SELinux for the `./docker-data/dms/config/:/tmp/docker-mailserver/` mount, append `-z` or `-Z`. Then configure the environment specific to the mail server by editing [`mailserver.env`][docs-environment], but keep in mind that:

Author: See f3b7fdc744a864a79f8ba9215b2ef101f0618282.

- [_only_ basic `VAR=VAL`][docker-compose-env-file] is supported (**do not** quote your values)
- variable substitution is **not** supported (e.g. :no_entry_sign: `OVERRIDE_HOSTNAME=$HOSTNAME.$DOMAINNAME` :no_entry_sign:)

[docs-environment]: ./config/environment.md
[docker-compose-env-file]: https://docs.docker.com/compose/env-file/

### Get Up and Running

!!! danger "Using the Correct Commands For Stopping and Starting DMS"

    **Use `docker compose up / down`, not `docker compose start / stop`**. Otherwise, the container is not properly destroyed and you may experience problems during startup because of inconsistent state.

    Using `Ctrl+C` **is not supported either**!

You are able to get a full overview of how the configuration works by running `docker run --rm ghcr.io/docker-mailserver/docker-mailserver:${TAG} setup help`, which provides you with all the information on configuration.

??? info "Usage of `setup.sh` when no DMS Container Is Running"

>     We encourage you to directly use `setup` inside the container (for instructions, see `docker exec -ti <CONTAINER NAME> setup help`). If you still want to use `setup.sh`, here's some information about it.

Author: > This is just repeating the same command as above almost, but with `docker exec` instead of `docker run`.

Indeed, but I want to make sure users understand that they actually do not really need `setup.sh` anymore :)

> I imagine most users would just use `docker exec` with `bash` to run multiple `setup ...` commands, or with `setup` instead of `bash` if automating through some scripts.

Probably 🤔

> `setup.sh` is mostly complimentary at this point, and not likely to be too useful for most users? (_beyond saving a few key strokes which an alias could do just as well_)

Yes, I think so too.

>     We encourage you to directly use `setup` inside the container (for instructions, see `docker exec -ti <CONTAINER NAME> setup help`). If you still want to use `setup.sh`, here's some information about it.

Reviewer: ```suggestion
    We encourage you to directly use `setup` inside the container (like shown above). If you still want to use `setup.sh`, here's some information about it.
```

```bash
docker exec -ti <CONTAINER NAME> setup alias add postmaster@example.com user@example.com
```

### DKIM Keys

You can (and you should) generate DKIM keys. We encourage you to [use Rspamd for DKIM signing][rspamd-dkim-signing]. If you still want to use OpenDKIM, you can use

``` BASH
docker exec -ti <CONTAINER NAME> setup config dkim help
```

to help you with generating keys.

In case you're using LDAP, the setup looks a bit different as you do not add user accounts directly. Postfix doesn't know your domain(s) and you need to provide it when configuring DKIM:

``` BASH
docker exec -ti <CONTAINER NAME> setup config dkim domain '<domain.tld>[,<domain2.tld>]'
```

When keys are generated, you can configure your DNS server by just pasting the content of `config/opendkim/keys/domain.tld/mail.txt` to [set up DKIM][dkim-signing-setup]. See the [documentation][docs-dkim-signing] for more details.

[dkim-signing-setup]: https://mxtoolbox.com/dmarc/dkim/setup/how-to-setup-dkim
[rspamd-dkim-signing]: ./config/security/rspamd.md#dkim-signing

> [docs-dkim-signing]: ./config/best-practices/dkim.md

Author: I like your suggestion! I actually wanted to revise the DKIM page, but didn't want to make this PR bigger than it already is - agreed!

> [docs-dkim-signing]: ./config/best-practices/dkim.md

Reviewer: > didn't want to make this PR bigger than it already is - agreed!

Yeah, lets not add anything new here haha, I am cutting it close on review deadline already :sweat_smile:

> [docs-dkim-signing]: ./config/best-practices/dkim.md

Author: Deffered to future PR (in the works).

        setup - 'docker-mailserver' Administration & Configuration script
    ...

    $ docker run --rm ghcr.io/docker-mailserver/docker-mailserver:latest setup help
    SETUP(1)

    NAME
        setup - 'docker-mailserver' Administration & Configuration script
    ...
    ```

On first start, you will need to add at least one email account (unless you're using LDAP). You have two minutes to do so, otherwise DMS will shutdown and restart. You can add accounts by running `docker exec -ti <CONTAINER NAME> setup email add user@example.com`. **That's it! It really is that easy**.

## Further Miscellaneous Steps

### Aliases

You should add at least one [alias][docs-aliases], the [_postmaster alias_][docs-env-postmaster]. This is a common convention, but not strictly required.

[docs-aliases]: ./config/user-management/aliases.md
[docs-env-postmaster]: ./config/environment.md#postmaster_address

```bash
docker exec -ti <CONTAINER NAME> setup alias add postmaster@example.com user@example.com
```

### DKIM Keys

> You can (_and you should_) generate DKIM keys. For more information:
> - DKIM [with OpenDKIM][docs-dkim-opendkim] (_enabled by default_)
> - DKIM [with Rspamd][docs-dkim-rspamd] (_when using `ENABLE_RSPAMD=1`_)

Reviewer: ```suggestion
You can (_and you should_) generate DKIM keys. For more information:

- DKIM [with OpenDKIM][docs-dkim-opendkim] (_enabled by default_)
- DKIM [with Rspamd][docs-dkim-rspamd] (_when using `ENABLE_RSPAMD=1`_)
```
